### ISSUE TITLE
feature: do not escape quotes in comments

### ISSUE THREAD

**Issue by jvanasco** [mozilla/bleach#604]
First, thanks for creating the 3.3.0 bugfix so quickly.  I looked at the timeline and it was pretty impressive.

The approach that landed has been a bit problematic for us - as we store metadata in comment tags.  Looking over the CVE and the approach, I'm not sure the single/double quotation marks were necessary.  The issue seems to be resolved when html brackets were escaped.  If my understanding is correct, allowing quotation marks in comments would largely make the  backwards incompatibility irrelevant as it would only apply to malformed html and attack vectors.

**Comment by willkg** [comment]
Can you include an example of the issue here? What does it mean to store metadata in the comment tags?

**Comment by willkg** [comment]
@jvanasco Can you include an example of the issue? What does it mean to store metadata in the comment tags?

**Comment by jvanasco** [comment]
In the most relavent example, upon submission of User Generated Content, we typically store some metadata about the submission itself within a comment tag in the UGC payload itself.

Attributes for comment tags are not typically supported by browsers - there is no reason to - but is allowed by the HTML spec and is often supported by parsing libraries.

Much of this is internal information, but one of the elements is typically the version of bleach used to sanitize the content and an internal identifier for the settings profile used:

    
    
This lets us use previously sanitized content when we detect a safe sanitizer version+profile match.  This helps us manage things a bit easier to re-sanitize on-demand, but only as necessary.

We also use the same strategy on editable user-generated content to trigger the safe rendering of approved embedded elements, this is essentially a variant of implementing oembed style content.  We package the "instructions" for the embed within a comment tag, so if there are any issues on rendering the fallback is not visible to users.  

    

In those cases, we just (i) regex out the comment nodes, (ii) process them into safe nodes, (iii) replace into the source node.

Bleach's approach to address the CVE included escaping (single and double) quotations, which meant we've needed to implement some heavy processing to reformat comment tags back into normal tags (i.e. tags with attributes) -- as they might already contain escaped quotes.

After reading the security tickets, CVE, and ancillary materials related to the CVE, escaping the quotations appears to be unnecessary. The proof-of-concept exploit for Bleach did not utilize quotations.  Most technical papers describing the general exploit technique do not utilize quotations.  I did find one example that used quotations, but they were unrelated to the actual exploit mechanism, and simply used to define a payload/variable used after the exploit has taken effect.  The exploit itself appears to be entirely caused by html brackets within a comment span not being escaped.  In Bleach's case, this vulnerability was because html5lib did not parse the malformed payload - caused by brackets - into compliant DOM. 

I've stared at the exploit examples and tried to make malformed payloads against earlier versions of bleach - I can't envision a way that escaping quotes adds any benefit once the malformed brackets are escaped.  Once those brackets are escaped, the quotes are simply text elements, which would otherwise be allowed.  

We handle this with a lot of pre/post regex processing -- which is fine.  Bleach is only one step in our content processing pipleline.

This is just a little annoying, and I'd prefer to not switch to another tag for storing metadata.  Being able to have attributes on a tag is really important to us - and the way this CVE fix was implemented really made the comment tag problematic
